staging/src/k8s.io/pod-security-admission/admission/admission.go

```diff
@@ -325,20 +347,33 @@ func (a *Admission) ValidatePod(ctx context.Context, attrs Attributes) *admissio
 		}
 		if !isSignificantPodUpdate(pod, oldPod) {
 			// Nothing we care about changed, so always allow the update.
-			return allowedResponse()
+			return sharedAllowedResponse()
 		}
 	}
 	return a.EvaluatePod(ctx, attrs.GetNamespace(), &pod.ObjectMeta, &pod.Spec, true)
 }
 
+// ValidatePodController evaluates a pod controller create or update request against the effective policy for the namespace.
+// The returned response may be shared between evaluations and must not be mutated.
 func (a *Admission) ValidatePodController(ctx context.Context, attrs Attributes) *admissionv1.AdmissionResponse {
 	// short-circuit on subresources
 	if attrs.GetSubresource() != "" {
-		return allowedResponse()
+		return sharedAllowedResponse()
 	}
 	// short-circuit on exempt namespaces and users
 	if a.exemptNamespace(attrs.GetNamespace()) || a.exemptUser(attrs.GetUserName()) {
-		return allowedResponse()
+		return sharedAllowedResponse()
+	}
+
+	// short-circuit on privileged audit+warn namespaces
+	namespace, err := a.NamespaceGetter.GetNamespace(ctx, attrs.GetNamespace())
+	if err != nil {
+		klog.ErrorS(err, "failed to fetch pod namespace", "namespace", attrs.GetNamespace())
+		return internalErrorResponse(fmt.Sprintf("failed to lookup namespace %s", attrs.GetNamespace()))
+	}
+	nsPolicy, _ := a.PolicyToEvaluate(namespace.Labels)
```

Comment:
We should just pass the parsed policy to `EvaluatePod` (instead of namespace) to avoid parsing it twice. The only problem is it makes it more complicated to record the error annotation... you could defer a function to add it to the response - just make sure it's not the shared response.

Reply:
saved ~5% CPU time in paths that hit both